Add xWaitForCertificateServices Resource - Fixes #73

PlagueHO opened the pull request:
This PR adds a new resource that waits for the specified CA to come online before continuing. It is intended to fix #73.

It also fixes issue #80 that was identified when I was creating/testing this resource.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xcertificate/81)
<!-- Reviewable:end -->

johlju commented:
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[ReadMe.md, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoawFpTCHiJHP_YOk82:-KoawFpTCHiJHP_YOk83:b-h4q1yi) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L18)):*
> ```Markdown
> - **xCertificateExport**: Used to export a certificate from a Windows certificate
>   store.
> - **xWaitForCA**: Used to wait for a CA to become available.
> ```

Could we write out CA to Certificate Authority, and be more specific that it is a Active Directory Certificate Service Certificate Authority? 

And all others throughout.

---

*[ReadMe.md, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koax5zcalEr8nttNtF_:-Koax5zcalEr8nttNtFa:bep7nly) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L18)):*
> ```Markdown
> - **xCertificateExport**: Used to export a certificate from a Windows certificate
>   store.
> - **xWaitForCA**: Used to wait for a CA to become available.
> ```

Suggestion (non-blocking): I would rather have seen this resource name would not have used abbreviation, but rather that it was named xWaitForCertificateAuthority, or xWaitForCertificateServices, because we cant wait for any CA, only ADCS CA's.

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koawo2In-Y76l6xoAuj:-Koawo2In-Y76l6xoAuk:b-1ckrsr) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
> ADCS CA

Should we write out to; Active Directory Certificate Service Certificate Authority

And all others throughout.

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoaxdfROgzt5Bjw4E8P:-KoawmER_Ze8DjwYzMcl:bxu8y5r) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
> ```Markdown
> ### xWaitForCA
> 
> - **`[String]` CAServerFQDN** (_Key_): The FQDN of the ADCS CA to wait for.
> ```

Can we remove the abbreviations in the parameter names?

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 29 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koay20ACCddAHVqkMwJ:-Koay20ACCddAHVqkMwK:b-pdv3kd) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L29)):*
> CA

Could we not use abbreviation? And all others throughout.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 32 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koay9_gMy7fvGIdu3Ve:-Koay9_gMy7fvGIdu3Vf:b2zm3m6) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L32)):*
> ADCS CA 

Could we not use abbreviation? And all others throughout.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 59 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoayXRsvkTu5QtGeMyU:-KoayXRsvkTu5QtGeMyV:b7v6fcv) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L59)):*
> $RetryIntervalSec

Can we change this to `RetryIntervalSeconds`?

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 66 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoayI-kJCik0k8ayed9:-KoayI-kJCik0k8ayedA:b3kvfwv) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L66)):*
> $ca

Could you change this variable name to something better, and not use abbreviation? 

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 124 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koaysq5gcmpymSVHzzA:-Koaysq5gcmpymSVHzzB:bnb4m7s) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L124)):*
> CheckingForCAStatusMessage

Abbreviation in these as well, can we change? 

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 183 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazMDFvDtc-56wE38-:-KoazMDFvDtc-56wE380:b-ybhvuw) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L183)):*
> ```PowerShell
>     param
>     (
>         [parameter(Mandatory = $true)]
> ```

Upper 'P'. Throughout.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 213 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-NZMrXmMlusPdZBa:-Kob-NZMrXmMlusPdZBb:bmpdg5h) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L213)):*
> ```PowerShell
>             Write-Verbose -Message ( @(
>                     "$($MyInvocation.MyCommand): "
>                     $($localizedData.CAOnlineMessage -f $ca)
> ```

Missing this one in the strings.psd1.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1, line 221 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-RkIxLZNkcgqrW3E:-Kob-RkIxLZNkcgqrW3F:bmpdg5h) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.psm1#L221)):*
> ```PowerShell
>     Write-Verbose -Message ( @(
>             "$($MyInvocation.MyCommand): "
>             $($localizedData.CAOfflineMessage -f $ca)
> ```

Missing this one in the strings.psd1.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.schema.mof, line 1 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazVVRqrQkIdeSThYU:-KoazVVSCV04V_B_jlgV:bbtb1ev) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.schema.mof#L1)):*
> ```
> 
> ```

Blank row here.

---

*[DSCResources/MSFT_xWaitForCA/en-us/MSFT_xWaitForCA.strings.psd1, line 3 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazY3s0s7w0b0G9gnS:-KoazY3s0s7w0b0G9gnT:bbtb1ev) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/en-us/MSFT_xWaitForCA.strings.psd1#L3)):*
> ```PowerShell
> ConvertFrom-StringData @'
>     GettingWaitForCAStatusMessage = Getting Wait for status of CA '{0}'.
> ```

Blank row here.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 17 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazkItbU_JaDdfRdWH:-KoazkIunMeMj_yhgsK7:b-jp32c3) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L17)):*
> ```PowerShell
>     (
>         [Parameter()]
>         [string[]]
> ```

System.String[]

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 22 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koazya6T3R882Zjnv8W:-Koazu0ZqGVzgrylqjdH:beb8cfi) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L22)):*
> ```PowerShell
>         [Parameter(Mandatory = $true)]
>         [ValidateNotNullorEmpty()]
>         [PSCredential]
> ```

[System.Management.Automation.PSCredential]

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 27 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob--kmf9DG92I6Vw51:-Kob--knEDhf8PrORNBW:b-iqa19c) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L27)):*
> ```PowerShell
> 
>     Import-DscResource -ModuleName xCertificate
>     Node 'localhost'
> ```

Blank row before this one.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 40 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-ca9ueJ-YxnSU0iF:-Kob-ca9ueJ-YxnSU0iG:b-ck4oj6) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L40)):*
> ```PowerShell
>             CAServerFQDN        = 'dc01.test.pha'
>             Subject             = 'foodomain.test.net'
>             KeyLength           = '1024'
> ```

Maybe promote key length of 2048 from now?

---

*[Modules/CertificateDsc.Common/CertificateDSc.Common.psm1, line 461 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob0MdbSgrpASkAKNdw:-Kob0MdbSgrpASkAKNdx:b-umqq2n) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Modules/CertificateDsc.Common/CertificateDSc.Common.psm1#L461)):*
> "-ping `"$CAServerFQDN\$CARootName`""

Maybe `('-ping "{0}\{1}"' -f $CAServerFQDN, $CARootName)`
Not sure it more readable. Non-blocking.

---

*[Tests/Unit/CertificateDsc.Common.Tests.ps1, line 871 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2J0H-h_lGCWfLQmD:-Kob2J0IIP9wIzPdoI_a:bawfdmn) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/CertificateDsc.Common.Tests.ps1#L871)):*
> ```PowerShell
>                 }
> 
>                 It 'Should call exepcted mocks' {
> ```

Typo 'expected'

---

*[Tests/Unit/MSFT_xWaitForCA.Tests.ps1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2pABfgcHiKMLICQP:-Kob2pABfgcHiKMLICQQ:b-amp0m7) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCA.Tests.ps1#L46)):*
> ```PowerShell
> 
>                 It 'Should return a hashtable' {
>                     ($result -is [hashtable]) | Should Be $true
> ```

$result | Should BeOfType [System.Collections.Hashtable]

---

*[Tests/Unit/MSFT_xWaitForCA.Tests.ps1, line 49 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2hM5RpBSaFOuC62z:-Kob2hM5RpBSaFOuC63-:ba8qvjy) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCA.Tests.ps1#L49)):*
> Should contain the input values

More clear? 'Should contain the same values passed as parameters'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcertificate/81)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Thanks for reviewing @Johlju ! I really appreciate it - I've made all the fixes I think (except changing the names of the two parameters - because I think they should be consistent with xCertReq).

I also ended up fixing up the style on all the examples, so there is an additional change thrown in.

I haven't changed the default key length in the xCertReq resource yet, but I've changed the examples to all be 2048 bits and also raised an issue to change the default key length to issue in xCertReq as well (I'll get that change made after this one is through).

---

Review status: 1 of 24 files reviewed at latest revision, 22 unresolved discussions.

---

*[ReadMe.md, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoawFpTCHiJHP_YOk82:-KofQ-Gzubj0BkGDY_9x:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L18)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could we write out CA to Certificate Authority, and be more specific that it is a Active Directory Certificate Service Certificate Authority? 

And all others throughout.
</blockquote></details>

Done.

---

*[ReadMe.md, line 18 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koax5zcalEr8nttNtF_:-KofQ-u5dLjy_-b-S7_h:b-30xosn) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L18)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Suggestion (non-blocking): I would rather have seen this resource name would not have used abbreviation, but rather that it was named xWaitForCertificateAuthority, or xWaitForCertificateServices, because we cant wait for any CA, only ADCS CA's.
</blockquote></details>

Done. Changed to xWaitForCertificateServices

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koawo2In-Y76l6xoAuj:-KofQ4dGVcmaLzerz3ls:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ADCS CA

Should we write out to; Active Directory Certificate Service Certificate Authority

And all others throughout.
</blockquote></details>

Done.

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoaxdfROgzt5Bjw4E8P:-KofQ6J3-pweErDMFlN8:b7w5d0e) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Can we remove the abbreviations in the parameter names?
</blockquote></details>

I'd prefer to leave the names as is so that they match the ones in xCertReq resource. Otherwise I think it would be pretty confusing to users. And changing xCertReq would be a breaking change. Is that OK? 

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 29 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koay20ACCddAHVqkMwJ:-KofQUfFKy8aVgHiDKA3:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L29)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> CA

Could we not use abbreviation? And all others throughout.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 32 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koay9_gMy7fvGIdu3Ve:-KofQVEw5w9QJxPn2qt-:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L32)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ADCS CA 

Could we not use abbreviation? And all others throughout.

</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 59 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoayXRsvkTu5QtGeMyU:-KofQX-bXmkt9BiPcxW8:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L59)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> $RetryIntervalSec

Can we change this to `RetryIntervalSeconds`?
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 66 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoayI-kJCik0k8ayed9:-KofRH6RF-GsIjCNzs09:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L66)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> $ca

Could you change this variable name to something better, and not use abbreviation? 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 124 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koaysq5gcmpymSVHzzA:-KofSZhbhRJJhtmVBMmo:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L124)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> CheckingForCAStatusMessage

Abbreviation in these as well, can we change? 
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 183 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazMDFvDtc-56wE38-:-KofUbLM58gxpnKOduAo:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L183)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Upper 'P'. Throughout.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 213 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-NZMrXmMlusPdZBa:-KofUcyeiphjlbwy0UZr:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L213)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Missing this one in the strings.psd1.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1, line 221 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-RkIxLZNkcgqrW3E:-KofUdh7uaFFstIMKjbz:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCertificateServices/MSFT_xWaitForCertificateServices.psm1#L221)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Missing this one in the strings.psd1.
</blockquote></details>

Done.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 17 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazkItbU_JaDdfRdWH:-KofUmzYcAtSal0ZbU2b:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L17)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

System.String[]
</blockquote></details>

Done.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 22 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Koazya6T3R882Zjnv8W:-KofUnNM81_9eQb0cKKg:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L22)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

[System.Management.Automation.PSCredential]
</blockquote></details>

Done.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 27 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob--kmf9DG92I6Vw51:-KofUoIlKA2C7DbYTYEx:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L27)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Blank row before this one.
</blockquote></details>

Done.

---

*[Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1, line 40 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob-ca9ueJ-YxnSU0iF:-KofUpRez8YDI_4rAtwD:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Examples/Resources/xWaitForCA/1-RequestSSLCertWithWait.ps1#L40)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe promote key length of 2048 from now?
</blockquote></details>

Done.

---

*[Modules/CertificateDsc.Common/CertificateDSc.Common.psm1, line 461 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob0MdbSgrpASkAKNdw:-KofV9MWmHbBQUKf0cTD:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Modules/CertificateDsc.Common/CertificateDSc.Common.psm1#L461)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> "-ping `"$CAServerFQDN\$CARootName`""

Maybe `('-ping "{0}\{1}"' -f $CAServerFQDN, $CARootName)`
Not sure it more readable. Non-blocking.
</blockquote></details>

Done.

---

*[Tests/Unit/CertificateDsc.Common.Tests.ps1, line 871 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2J0H-h_lGCWfLQmD:-KofVKxAi-k--vh1r1nd:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/CertificateDsc.Common.Tests.ps1#L871)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Typo 'expected'
</blockquote></details>

Done.

---

*[Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2pABfgcHiKMLICQP:-KofVTDrFzCTJ8RYjV5d:b-fwums3) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1#L46)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

$result | Should BeOfType [System.Collections.Hashtable]
</blockquote></details>

Done. Did global search/replace which corrected it in other tests.

---

*[Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1, line 49 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2hM5RpBSaFOuC62z:-KofVk_kczC6FQGeHwF3:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1#L49)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> Should contain the input values

More clear? 'Should contain the same values passed as parameters'
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.schema.mof, line 1 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazVVRqrQkIdeSThYU:-KofVlgZ9FcnOhJkYjRr:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/MSFT_xWaitForCA.schema.mof#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Blank row here.
</blockquote></details>

Done.

---

*[DSCResources/MSFT_xWaitForCA/en-us/MSFT_xWaitForCA.strings.psd1, line 3 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoazY3s0s7w0b0G9gnS:-KofVmF8JuuJhFMcR0X1:b-896fix) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/DSCResources/MSFT_xWaitForCA/en-us/MSFT_xWaitForCA.strings.psd1#L3)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Blank row here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcertificate/81#-:-KofS5FR6RPbxzAijwKc:b-itg7i7)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 29 of 29 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoaxdfROgzt5Bjw4E8P:-KogEFydpoTWrk3JCF1T:b-7l18t5) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

I'd prefer to leave the names as is so that they match the ones in xCertReq resource. Otherwise I think it would be pretty confusing to users. And changing xCertReq would be a breaking change. Is that OK? 
</blockquote></details>

Okay, let's leave them. That could be changed throughout (all resources) when this module becomes HQRM (changes name).

---

*[ReadMe.md, line 143 at r2](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KogE0FirWzttqMeIfoX:-KogE0FirWzttqMeIfoY:bw5uryk) ([raw file](https://github.com/powershell/xcertificate/blob/4adbe616f03641b1fd76ffbdfa513845f2ce2219/ReadMe.md#L143)):*
> ```Markdown
> - **`[Uint32]` RetryIntervalSeconds**: Specifies the number of seconds to wait for
>   the Active Directory Certificate Service Certificate Authority to become
>   available. Defaults to 10.
> ```

Maybe: 'Defaults to 10 seconds.'

---

*[ReadMe.md, line 146 at r2](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KogE4gzaXaJbB5y4gKc:-KogE4gzaXaJbB5y4gKd:b-vk6nte) ([raw file](https://github.com/powershell/xcertificate/blob/4adbe616f03641b1fd76ffbdfa513845f2ce2219/ReadMe.md#L146)):*
> ```Markdown
> - **`[Uint32]` RetryCount**: The number of times to loop the retry interval while
>   waiting for the Active Directory Certificate Service Certificate Authority.
>   Optional. Defaults to 60.
> ```

Maybe: 'Defaults to 60 retries.'

---

*[Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2pABfgcHiKMLICQP:-KogGQaWCkxs39AOEnMs:bn6gorm) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1#L46)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Done. Did global search/replace which corrected it in other tests.
</blockquote></details>

You missed this one? It does not use `BeOfType` (and not in other tests either).
It's okay as-is. So making this a non-blocking comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcertificate/81)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Thanks for the review @johlju. I think I've fixed everything. I included a couple of other minor style fixes here and there where I saw them. I'll actually do a full HQRM run through of this module when I move it over to auto-documentation in the next week. So any other style fixes will be made then (it'll be a big PR I imagine but I'll try and break it up into smaller chunks like I'm doing with xNetworking).

---

Review status: 16 of 26 files reviewed at latest revision, 2 unresolved discussions.

---

*[ReadMe.md, line 136 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KoaxdfROgzt5Bjw4E8P:-KoiJ-7c2AGphZXMJoEZ:b-a3z1fg) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/ReadMe.md#L136)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Okay, let's leave them. That could be changed throughout (all resources) when this module becomes HQRM (changes name).
</blockquote></details>

Cool. Will do this as part of HQRM (if HQRM is still going to happen :grin:) - I've logged an issue to do this.

---

*[ReadMe.md, line 143 at r2](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KogE0FirWzttqMeIfoX:-KoiL1qHsc2fnol_15OP:b-r2yxmw) ([raw file](https://github.com/powershell/xcertificate/blob/4adbe616f03641b1fd76ffbdfa513845f2ce2219/ReadMe.md#L143)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe: 'Defaults to 10 seconds.'
</blockquote></details>

Done. And in the MOF/PSM1 as well.

---

*[ReadMe.md, line 146 at r2](https://reviewable.io:443/reviews/powershell/xcertificate/81#-KogE4gzaXaJbB5y4gKc:-KoiL1bSxSqRulpW2aFX:b-r2yxmw) ([raw file](https://github.com/powershell/xcertificate/blob/4adbe616f03641b1fd76ffbdfa513845f2ce2219/ReadMe.md#L146)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe: 'Defaults to 60 retries.'
</blockquote></details>

Done. And in the MOF/PSM1 as well.

---

*[Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1, line 46 at r1](https://reviewable.io:443/reviews/powershell/xcertificate/81#-Kob2pABfgcHiKMLICQP:-KoiLkiMKJOX73gE2aFR:b-bz6822) ([raw file](https://github.com/powershell/xcertificate/blob/5861bb0f3d4f61afa9bfa9f8803ae45916f0bd7e/Tests/Unit/MSFT_xWaitForCertificateServices.Tests.ps1#L46)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

You missed this one? It does not use `BeOfType` (and not in other tests either).
It's okay as-is. So making this a non-blocking comment.
</blockquote></details>

Oops! Good catch (I misunderstood the suggestion). Fixed now (and all others).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcertificate/81#-:-KoiLt_ctBHMIg1aUOnl:ba2dexh)*
<!-- Sent from Reviewable.io -->

johlju commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

You are welcome :smile: 

---

Reviewed 2 of 29 files at r2, 10 of 10 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xcertificate/81#-:-KoiOYfducF0e95M_MSQ:b-55mnxs)*
<!-- Sent from Reviewable.io -->
